testdata/p4_16_samples/psa-portid-using-newtype2.p4

```diff
@@ -0,0 +1,293 @@
+// headers in packets from the CPU port.
+@p4runtime_translate("port")
+newtype bit<32> PortIdInHeader_t;
+
+// used to cast between PortId_t and PortIdInHeader_t the programmer needs to be
+// aware that he may be exposing non-translated values to the control-plane,
+// unless he use @p4runtime_translate appropriately
+
+// Note: all of these casts are necessary with the current newtype
+// code.  Eliminating any one of them results in a compiler error when
+// using the macro as done later in this program.  The nice thing is
+// that this macro could be added to psa.p4, and most people won't
+// need to see it.
+
+#define PSA_PORT_HEADER_TO_ID(p) ((PortId_t) (PortIdUInt_t) (bit<32>) (p))
+#define PSA_PORT_ID_TO_HEADER(p) ((PortIdInHeader_t) (bit<32>) (PortIdUInt_t) (p))
+
+// Included here simply so I can use type PortId_t in my
+// locally-modified version of v1model.p4, called v1model-newtype.p4.
+#include "v1model-newtype.p4"
+
+
+header ethernet_t {
+    bit<48> dstAddr;
+    bit<48> srcAddr;
+    bit<16> etherType;
+}
+
+header ipv4_t {
+    bit<4>  version;
+    bit<4>  ihl;
+    bit<8>  diffserv;
+    bit<16> totalLen;
+    bit<16> identification;
+    bit<3>  flags;
+    bit<13> fragOffset;
+    bit<8>  ttl;
+    bit<8>  protocol;
+    bit<16> hdrChecksum;
+    bit<32> srcAddr;
+    bit<32> dstAddr;
+}
+
+
+// From fabric.p4
+@controller_header("packet_in")
+header packet_in_header_t {
+    // Translation will occur in agent software for this field,
+    // because it is inside of a controller_header, and because
+    // PortIdInHeader_t is defined as a newtype with a
+    // @p4runtime_translate("port") annotation.  No such annotation is
+    // needed here.
+    PortIdInHeader_t ingress_port;
+}
+
+@controller_header("packet_out")
+header packet_out_header_t {
+    // No annotation needed here to get agent translation of
+    // egress_port, for same reason as in header packet_in_header_t
+    // above.
+    PortIdInHeader_t egress_port;
+}
+
+
+struct parsed_headers_t {
+    packet_in_header_t packet_in;
+    packet_out_header_t packet_out;
+    ethernet_t ethernet;
+    ipv4_t ipv4;
+}
+
+struct fabric_metadata_t {
+}
+
+action drop() {
+    mark_to_drop();
+}
+
+action nop() {
+}
+
+parser FabricParser(packet_in packet,
+                    out parsed_headers_t hdr,
+                    inout fabric_metadata_t meta,
+                    inout standard_metadata_t standard_metadata)
+{
+    state start {
+        transition accept;
+    }
+}
+
+control PacketIoIngress(inout parsed_headers_t hdr,
+                        inout fabric_metadata_t fabric_metadata,
+                        inout standard_metadata_t standard_metadata)
+{
+    apply {
+        if (hdr.packet_out.isValid()) {
+            // I know what I'm doing, I requested translation on
+            // packet_out.egress_port
+            standard_metadata.egress_spec =
+                PSA_PORT_HEADER_TO_ID(hdr.packet_out.egress_port);
```

Comment:
Given the comment above, I think that the simple cast should be enough and that this will be even more readable than the macro.
```
standard_metada.egress_spec = (PortId_t)hdr.packet_out.egress_port;
```

Reply:
Again, I don't have any good ideas on how to change the compiler so that one cast is enough, but the macro definition as proposed does do what is needed here.